src/allmydata/test/test_storage_client.py

```diff
@@ -739,3 +735,64 @@ def add_one_server(x):
 
         yield done
         self.assertTrue(done.called)
+
+
+class PickHTTPServerTests(unittest.SynchronousTestCase):
+    """Tests for ``_pick_a_http_server``."""
+
+    def loop_until_result(self, url_to_results: dict[DecodedURL, list[tuple[float, Union[Exception, Any]]]]) -> tuple[int, DecodedURL]:
+        """
+        Given mapping of URLs to list of (delay, result), return the URL of the
+        first selected server.
+        """
+        clock = Clock()
+
+        def request(reactor, url):
+            delay, value = url_to_results[url].pop(0)
+            result = Deferred()
+            def add_result_value():
+                if isinstance(value, Exception):
+                    result.errback(value)
+                else:
+                    result.callback(value)
+            reactor.callLater(delay, add_result_value)
+            return result
+
+        iterations = 0
+        while True:
+            iterations += 1
+            d = _pick_a_http_server(clock, list(url_to_results.keys()), request)
+            for i in range(100):
+                clock.advance(0.1)
+            result = self.successResultOf(d)
+            if result is not None:
+                return iterations, result
+
+    def test_first_successful_connect_is_picked(self):
+        """
+        Given multiple good URLs, the first one that connects is chosen.
+        """
+        earliest_url = DecodedURL.from_text("http://a")
+        latest_url = DecodedURL.from_text("http://b")
+        iterations, result = self.loop_until_result({
+            latest_url: [(2, None)],
+            earliest_url: [(1, None)]
+        })
+        self.assertEqual(iterations, 1)
+        self.assertEqual(result, earliest_url)
+
+    def test_failures_are_retried(self):
+        """
+        If the initial requests all fail, ``_pick_a_http_server`` keeps trying
+        until success.
```

Comment:
It looks like this ended up being more of a test for `loop_until_result` than `_pick_a_http_server`, I think.  (Noting because the `race` I put in that Twisted PR does not have this behavior).

Reply:
Mmm. Yeah the loop used to be in `_pick_a_http_server` and had to get moved out for reasons.